### PR TITLE
rust/sip: register parser for tcp v4

### PR DESCRIPTION
**Pull request by glongo**
Make sure these boxes are signed before submitting your Pull Request -- thank you.

- [x] I have read the contributing guide lines at https://suricata.readthedocs.io/en/latest/devguide/codebase/contributing/contribution-process.html
- [x] I have signed the Open Information Security Foundation contribution agreement at https://suricata.io/about/contribution-agreement/
- [x] I have updated the user guide (in doc/userguide/) to reflect the changes made (if applicable)

Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket:
https://redmine.openinfosecfoundation.org/issues/3351

Previous pr: #8725 

Describe changes:
- Loop over input data when parsing request/response

suricata-verify-pr: 1171

### CODE REVIEW

File: src/output-json-sip.c [hunk 1]
```diff
@@ -65,7 +65,7 @@ static int JsonSIPLogger(ThreadVars *tv, void *thread_data,
     SIPTransaction *siptx = tx;
     OutputJsonThreadCtx *thread = thread_data;
 
-    JsonBuilder *js = CreateEveHeader((Packet *)p, LOG_DIR_PACKET, "sip", NULL, thread->ctx);
+    JsonBuilder *js = CreateEveHeader((Packet *)p, LOG_DIR_FLOW, "sip", NULL, thread->ctx);
```
Comment (catenacyber) [inline src/output-json-sip.c:68]: What does this difference make ? For UDP and for TCP ?

File: rust/src/sip/sip.rs [hunk 2]
```diff
@@ -325,6 +469,26 @@ pub unsafe extern "C" fn rs_sip_parse_request(
...
+#[no_mangle]
+pub unsafe extern "C" fn rs_sip_parse_request_tcp(
+    flow: *const core::Flow,
+    state: *mut std::os::raw::c_void,
+    pstate: *mut std::os::raw::c_void,
+    stream_slice: StreamSlice,
+    _data: *const std::os::raw::c_void,
+) -> AppLayerResult {
+    if stream_slice.is_empty() {
+        if AppLayerParserStateIssetFlag(pstate, APP_LAYER_PARSER_EOF_TS) > 0 {
```
Comment (catenacyber) [inline rust/src/sip/sip.rs:481]: Why is not this block in `rs_sip_parse_response_tcp` as well ?

File: rust/src/sip/sip.rs [hunk 3]
```diff
@@ -298,6 +392,31 @@ pub unsafe extern "C" fn rs_sip_probing_parser_ts(
     return ALPROTO_UNKNOWN;
 }
 
+#[no_mangle]
+pub unsafe extern "C" fn rs_sip_probing_parser_tcp_ts(
```
Comment (catenacyber) [inline rust/src/sip/sip.rs:396]: Could we rather have pattern matching instead of probing parser for SIP ?

File: rust/src/sip/sip.rs [hunk 4]
```diff
@@ -173,17 +224,60 @@ impl SIPState {
...
-    pub fn new(id: u64) -> SIPTransaction {
+    pub fn new(id: u64, direction: crate::core::Direction) -> SIPTransaction {
         SIPTransaction {
             id,
             request: None,
             response: None,
             request_line: None,
             response_line: None,
-            tx_data: applayer::AppLayerTxData::new(),
+            tx_data: applayer::AppLayerTxData::for_direction(direction),
```
Comment (catenacyber) [inline rust/src/sip/sip.rs:280]: This change for unidirectional transactions may deserve its own commit

File: rust/src/sip/sip.rs [hunk 5]
```diff
@@ -173,17 +224,60 @@ impl SIPState {
             }
         }
     }
+
+    fn parse_response_tcp(&mut self, flow: *const core::Flow, stream_slice: StreamSlice) -> AppLayerResult {
+        let input = stream_slice.as_slice();
+        if input.is_empty() {
+            return AppLayerResult::ok();
+        }
+        let _pdu = Frame::new(
```
Comment (catenacyber) [inline rust/src/sip/sip.rs:233]: I think this not a PDU, the PDU is the part of the stream slice corresponding to the `response` a few lines below

File: rust/src/sip/parser.rs [hunk 6]
```diff
@@ -65,7 +65,7 @@ pub enum Method {
 
 #[inline]
 fn is_token_char(b: u8) -> bool {
-    is_alphanumeric(b) || b"!%'*+-._`".contains(&b)
+    is_alphanumeric(b) || b"!%'*+-._`;=@".contains(&b)
```
Comment (catenacyber) [inline rust/src/sip/parser.rs:68]: Why these new characters ?

Comment (glongo) [reply rust/src/sip/parser.rs:68]: Because they are valid chars for the URI field, please look at the example below:
`REGISTER sip:192.168.43.100;transport=TCP SIP/2.0`

File: rust/src/sip/sip.rs (hunk 3 above)
Comment (glongo) [reply rust/src/sip/sip.rs:396]: I think there is already a ticket for that: https://redmine.openinfosecfoundation.org/issues/5047

May it be the next step after that this PR will be merged?

File: rust/src/sip/parser.rs (hunk 6 above)
Comment (catenacyber) [reply rust/src/sip/parser.rs:68]: Only for TCP ?
Is there a RFC referencing this ?

Comment (glongo) [reply rust/src/sip/parser.rs:68]: No, TCP and UDP.
Yes, it's defined in SIP RFC page 222-223: https://www.rfc-editor.org/rfc/rfc3261

> 
> SIP-URI          =  "sip:" [ userinfo ] hostport
>                     uri-parameters [ headers ]
> userinfo         =  ( user / telephone-subscriber ) [ ":" password ] "@"
> user             =  1*( unreserved / escaped / user-unreserved )
> user-unreserved  =  "&" / "=" / "+" / "$" / "," / ";" / "?" / "/"
> password         =  *( unreserved / escaped /
>                     "&" / "=" / "+" / "$" / "," )
> hostport         =  host [ ":" port ]
> host             =  hostname / IPv4address / IPv6reference
> uri-parameters    =  *( ";" uri-parameter)
> uri-parameter     =  transport-param / user-param / method-param
>                      / ttl-param / maddr-param / lr-param / other-param
> transport-param   =  "transport="
>                      ( "udp" / "tcp" / "sctp" / "tls"
>                      / other-transport)
> ...
>

Comment (glongo) [reply rust/src/sip/parser.rs:68]: No, because `:` is not a valid token character, and then it's included here:
```
#[inline]
fn is_request_uri_char(b: u8) -> bool {
    is_alphanumeric(b) || is_token_char(b) || b"~#@:".contains(&b)
}
```
and token is defined in RFC3261 as
```
token       =  1*(alphanum / "-" / "." / "!" / "%" / "*"
                     / "_" / "+" / "`" / "'" / "~" )
```

Looking better at it, makes more sense to add the new chars in `is_request_uri_char` and not in `is_token_char`.

Comment (catenacyber) [reply rust/src/sip/parser.rs:68]: Thanks. These look rather uri than token indeed.
And token seems to be missing `~`

**Review by catenacyber — CHANGES_REQUESTED** [review]
Thanks for your work on this.
I guess the main improvement to do is to split the commit for each logical unit

**Comment by glongo** [comment]
Replaced by https://github.com/OISF/suricata/pull/9321.
